lib/semantics/mod-file.cc

```diff
@@ -0,0 +1,356 @@
+// Copyright (c) 2018, NVIDIA CORPORATION.  All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "mod-file.h"
+#include "scope.h"
+#include "symbol.h"
+#include "../parser/message.h"
+#include <algorithm>
+#include <cerrno>
+#include <fstream>
+#include <functional>
+#include <ostream>
+#include <set>
+#include <sstream>
+#include <vector>
+
+namespace Fortran::semantics {
+
+using namespace parser::literals;
+
+class ModFileWriter {
+public:
+  // The initial characters of a file that identify it as a .mod file.
+  static constexpr auto magic{"!mod$"};
+  static constexpr auto extension{".mod"};
+
+  // The .mod file format version number.
+  void set_version(int version) { version_ = version; }
```

Comment:
These might be better expressed as arguments to a constructor.

Reply:
Yes, I think we should revisit the API when we know more about how it will be used.